client/CPlayerInterface.cpp

```diff
@@ -2813,3 +2814,83 @@ void CPlayerInterface::showWorldViewEx(const std::vector<ObjectPosInfo>& objectP
 
 	viewWorldMap();
 }
+
+void CPlayerInterface::updateAmbientSounds()
+{
+	std::unordered_set<int3, ShashInt3> tiles;
+	int3 pos = currentSelection->getSightCenter();
+	cb->getVisibleTilesInRange(tiles, pos, 3, 2);
+	std::map<ObjectInstanceID, ui32> currentObjects;
+
+
+	for(int3 tile : tiles)
+	{
+		ui32 volume = 100;
+		switch(static_cast<int>(pos.chebdist2d(tile)))
+		{
+		case 1:
+			volume = 90;
+			break;
+		case 2:
+			volume = 60;
+			break;
+		case 3:
+			volume = 30;
+			break;
+		}
+		volume = (CCS->soundh->getVolume() / 100.0) * volume;
+
+		auto updateObjects = [&](ObjectInstanceID id) -> void
+		{
+			if(vstd::contains(currentObjects, id))
+			{
+				if(volume > currentObjects[id])
+					currentObjects[id] = volume;
+			}
+			else
+				currentObjects.insert(std::make_pair(id, volume));
+		};
+
+		auto tt = cb->getTile(tile);
+		if(tt->isWater())
+			updateObjects(ObjectInstanceID());
+
+		auto obj = tt->topVisitableObj(pos == tile);
+		if(obj && obj->ID != Obj::HERO)
+			updateObjects(obj->id);

```

Comment:
id is unique identifier of objetc instance. Shouldn't sounds be differentiated by <ID, subID> pair?